Add feature to reload apache service when content of ssl files has changed

timdeluxe opened the pull request:
This fixes issue MODULES-10847, which we also wanted to solve.

The solution with those helper files seems a bit odd, but the original ssl files could be managed by any means outside of the apache module or even outside of puppet, so this was the only way i could image to achieve it.

(Slightly offtopic: Let me add that this module gave me a love/hate relationship with spec/acceptance tests. First i blamed wrong tests to be the cause for failing, but in the end it was my own stupid code. Finally i could manage getting tests run green at my desktop - hope they will do in the modules CI setup as well.)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/puppetlabs/puppetlabs-apache?pullRequest=2157) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/puppetlabs/puppetlabs-apache?pullRequest=2157) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: timdeluxe<br/>:x: ThomasMinor<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/puppetlabs/puppetlabs-apache?pullRequest=2157) it.</sub>

timdeluxe commented:
Ah well, so many hours with tests and i oversaw the basics... 🙈 Please apologize!
Will fix and squash/rebase. No need to comment about this.

However if you have feedback to the feature itself, let me know.

ekohl commented on REFERENCE.md:
```diff
@@ -145,26 +179,92 @@ Apache server's or a virtual host's listening address and port.
-* [`apache::pw_hash`](#apachepw_hash): Hashes a password in a format suitable for htpasswd files read by apache.
-* [`apache_pw_hash`](#apache_pw_hash): DEPRECATED.  Use the namespaced function [`apache::pw_hash`](#apachepw_hash) instead.
-* [`bool2httpd`](#bool2httpd): DEPRECATED.  Use the namespaced function [`apache::bool2httpd`](#apachebool2httpd) instead.
+* [`stdlib::ensure`](#stdlibensure): function to cast ensure parameter to resource specific value
```
This looks wrong. As if it created a reference for all fixtures as well.

timdeluxe replied:
Correct, my fault. Fixed it in the new push

ekohl commented on manifests/mod/ssl.pp:
```diff
@@ -174,6 +178,46 @@
     include apache::mod::socache_shmcb
   }
 
+  file { $apache::params::puppet_ssl_dir:
```
Why refer to `apache::params` instead of `apache`?

timdeluxe replied:
Because it is defined only there? It is NOT a parameter in init.pp.
Maybe i misunderstand something about variable scoping and you can explain?

ekohl replied:
Technically `apache` inherits from `apache::params` so you can use `$apache::puppet_ssl_dir`. Then in the future you can easily make it a parameter.

I've seen that this confuses people and isn't really obvious to people.

timdeluxe replied:
I am still not convinced. I've seen a lot of other places, where apache::params::* variables are used.
It still can be changed, when it shall be come a parameter of the init class.
It would rather confuse me the other way around if i would use $apache::puppet_ssl_dir and don't have a parameter for that. I don't see the need for a parameter.

DavidS replied:
I don't have an opinion on which direction to go here, but do understand that many parts of this module are older than what anyone would call "best practice" today.

timdeluxe replied:
@DavidS Well, i think we then need a maintainer of this module to decide.

DavidS replied:
When in doubt, I go with what's there in the codebase and the opinion of them who does the work, unless there is a critical overriding factor.

ekohl replied:
>  It would rather confuse me the other way around if i would use $apache::puppet_ssl_dir and don't have a parameter for that. I don't see the need for a parameter.

That's what I've also seen from others. They assume it doesn't work because there's no parameter and forget the inheritance. Implementation wise it doesn't make a difference but code clarity would be a good reason to use `::params::`. I just wanted to highlight it is an option.

timdeluxe replied:
Thanks for highlighting. I think i leave it as it is, unless the maintainers of the module want a change.

ekohl commented on manifests/init.pp:
```diff
@@ -91,6 +91,9 @@
 #   this parameter with your SSL key's location before deploying this server in a production 
 #   environment.
 #
+# @param ssl_reload_on_change
```
This doesn't match the parameter name.

timdeluxe replied:
Correct, my fault, fixed.

ekohl commented on manifests/mod/ssl.pp:
```diff
@@ -97,6 +100,7 @@
   Optional[String] $stapling_cache                          = undef,
   Optional[Boolean] $ssl_stapling_return_errors             = undef,
   $ssl_mutex                                                = undef,
+  Boolean $ssl_reload_on_change                             = false,
```
Should this default to undef and then use `pick($ssl_reload_on_change, $apache::ssl_reload_on_change)`?

timdeluxe replied:
No, that makes no sense, see also our further discussions

ekohl reviewed (commented):
I personally like a minimal impact on environments. Can it only manage the directory if the feature is enabled? Perhaps even purge the whole directory if it's disabled?

timdeluxe commented:
> I personally like a minimal impact on environments. Can it only manage the directory if the feature is enabled? Perhaps even purge the whole directory if it's disabled?

@ekohl Before i answer to each of the code remarks, lets discuss this first, because those code remark discussions depend on that:

I struggled with exactly that question, but the reload thing should be a per-vhost-based thing. One should be able to have apache reloaded if the ssl files change for a vhost, but not for the other. At least that is my opinion. I need to manage the folder in ssl.pp then, because vhost.pp is a define and managing the folder there would lead to duplicate declarations. Also even if the reload is deactivated for default vhost and for the mod::ssl module, it still can be activated for a vhost and then the folder is required.

Or do we say that reload thing is a general setting and can't be changed on a per vhost base? (In that case your requested changes could be implemented)

ekohl commented:
You're right that we should take it to a higher design level.

So we have a few options. The current one obviously. We can also introduce a global parameter.

Another is introducing a private class (like `apache::mod::ssl::reload`) that does the directory setup. Then in `apache::vhost` you can include it if it does need the reload.

The benefit is that `mod_ssl` doesn't need any modification. It also keeps it "lazy" in that you only get the additional directories if you do use the feature. The downside is that you can't purge/clean it up.

timdeluxe commented:
A private class is indeed a good idea, will try that, thanks!

However mod_ssl does need at least the file resources for the ssl files, because one could work with apache without using vhosts at all (extremly rare, but the module is widely spread, so we should not forget those people).

timdeluxe commented:
On real world usage of my branch i found two further issues, which i now fixed: The puppet_ssl_dir was managed before apache was even installed and resulted in missing httpd_dir, so management of puppet_ssl_dir failed. Also even if the default-ssl vhost was set to absent, it still managed its certificate copy - this is also now fixed.

This PR still says @ekohl requested 1 change, however i can only find resolved conversations. Can someone help me?

ekohl commented on REFERENCE.md:
```diff
@@ -145,27 +146,6 @@ Apache server's or a virtual host's listening address and port.
 * `apache::peruser::processor`: Enables the `Peruser` module for FreeBSD only.
 * `apache::security::rule_link`: Links the activated_rules from `apache::mod::security` to the respective CRS rules on disk.
 
-### Resource types
```
These shouldn't be deleted because they're still present. Any idea why it's doing this?

timdeluxe replied:
Because i used puppet strings wrongly due to fixtures, which aren't ignored by it. I think i managed it now to have a correct reference

david22swan commented:
@timdeluxe As stated above by @ekohl it looks like some of the reference has been deleted by your changes.
Could you get that added back for us? Otherwise we won't be able to merge.

timdeluxe commented:
@david22swan Sorry, missed that. Is fixed now.

bastelfreak commented on manifests/init.pp:
```diff
@@ -472,6 +475,7 @@
   $default_ssl_crl_path                                                 = undef,
   $default_ssl_crl                                                      = undef,
   $default_ssl_crl_check                                                = undef,
+  $default_ssl_reload_on_change                                         = false,
```
could you please add a datatype for the new parameter?
```suggestion
  Boolean $default_ssl_reload_on_change                                 = false,
```

timdeluxe replied:
@bastelfreak Done.

david22swan commented:
@timdeluxe Apologies, for that to disappear @ekohl must approve the PR.

ekohl commented on manifests/vhost.pp:
```diff
@@ -2707,7 +2711,64 @@
+        $_ssl_cert_copy = regsubst($ssl_cert, '/', '_', 'G')
+        file { "${filename}${_ssl_cert_copy}":
+          path    => "${apache::params::puppet_ssl_dir}/${filename}${_ssl_cert_copy}",
+          source  => "file://${ssl_cert}",
```
Does this automatically create a dependency on that file or should it have something like:
```puppet
File <| title == $ssl_cert |> -> File["${filename}${_ssl_cert_copy}"]
```

timdeluxe replied:
We should avoid that and thats the whole point of the way i implemented it: We don't know how the cert file is managed. It may be a file resource in puppet, but does not have to be it. It could be managed completely outside of puppet or via another resource type.
If we would be sure, that the cert file is always a file resource, we could just do a notify dependency from the File to the apache::service and that would be all, but we shouldn't do that.
And yes, that "file://" locator creates a dependency: Once the content of the source file changes, also the "cert copy" file change, resulting into a notify into the apache::service class. This way it works independently of how the real cert files are managed.

ekohl commented on manifests/vhost.pp:
```diff
@@ -2707,7 +2711,64 @@
+    if $ssl_reload_on_change {
+      if $ssl_cert {
+        include apache::mod::ssl::reload
+        $_ssl_cert_copy = regsubst($ssl_cert, '/', '_', 'G')
+        file { "${filename}${_ssl_cert_copy}":
```
There's a lot of duplication for this. Can't this be captured in a define somehow and reused? Perhaps a loop?

```puppet
[$ssl_cert, $ssl_key, $ssl_ca, $ssl_chain, $ssl_crl].each |$ssl_file| {
  if $ssl_file {
    include apache::mod::ssl::reload
    $_ssl_file_copy = regsubst($ssl_file, '/', '_', 'G')
    file { "${filename}${_ssl_file_copy}":
      path    => "${apache::params::puppet_ssl_dir}/${filename}${_ssl_file_copy}",
      source  => "file://${ssl_file}",
      mode    => '0640',
      seltype => 'cert_t',
      notify  => Class['apache::service'],
    }
  }
} 
```

timdeluxe replied:
Thats actually a good idea, will implement that.

timdeluxe replied:
Did the same in apache::mod::ssl

timdeluxe commented on manifests/vhost.pp:
```diff
@@ -2707,7 +2711,64 @@
   # - $ssl_stapling
   # - $apache_version
-  if $ssl {
+  if $ssl and $ensure == 'present' {
+    if $ssl_reload_on_change {
```
No it can't - see line 2788, where the concat::fragment is located, which must be handled even if the reload feature is not used. Maybe it is clearer when i shift the concat::fragment higher and the reload stuff after that?

ekohl replied:
I do think that would be easier to follow. Long deeply nested structures aren't that clear. Perhaps it's also not needed if https://github.com/puppetlabs/puppetlabs-apache/pull/2157#discussion_r663036253 works and can be implemented.

timdeluxe replied:
Yes, you are right, but i will do the shift anyway even with the proposal of the other comment, because i find that also better.

ekohl commented on manifests/mod/ssl.pp:
```diff
@@ -62,6 +62,9 @@
 #   - Debian/Ubuntu + Apache < 2.4: 'file:${APACHE_RUN_DIR}/ssl_mutex'.
 #
+# @param ssl_reload_on_change
+#   Enable reloading of apache if the content of ssl files have changed.
```
Perhaps this should clarify that it's only the SSL files that are configured here in mod_ssl, not those in vhosts.

ekohl commented on manifests/init.pp:
```diff
@@ -91,6 +91,9 @@
 #   environment.
 #
+# @param default_ssl_reload_on_change
+#   Enable reloading of apache if the content of ssl files have changed.
```
I think this should clarify this **only** sets the default value for vhosts. If you don't have any SSL vhosts or use `mod_ssl` to configure the SSL certificates, it does nothing.

ekohl commented on manifests/vhost.pp:
```diff
@@ -2707,12 +2711,27 @@
   # - $ssl_stapling
   # - $apache_version
-  if $ssl {
+  if $ssl and $ensure == 'present' {
```
Should it actually ensure the old certificate copies are absent if `$ensure` is `absent`?

timdeluxe replied:
Not needed, the puppet_ssl_dir is managed with purge => true, so old copies will be cleaned up by that.

ekohl commented on manifests/mod/ssl.pp:
```diff
@@ -174,6 +178,22 @@
+        file { $_ssl_file_copy:
+          path    => "${apache::params::puppet_ssl_dir}/${_ssl_file_copy}",
+          source  => "file://${ssl_file}",
+          mode    => '0640',
```
Should the owner and group also be managed? (same question for the vhost)

ekohl commented on manifests/vhost.pp:
```diff
@@ -2707,12 +2711,27 @@
       order   => 230,
       content => template('apache/vhost/_ssl.erb'),
     }
+    if $ssl_reload_on_change {
```
Nit: I like the visual separation
```suggestion

    if $ssl_reload_on_change {
```

ekohl reviewed (commented):
Thanks! This is much easier to follow.

ekohl commented on manifests/init.pp:
```diff
@@ -91,6 +91,12 @@
 #   this parameter with your SSL key's location before deploying this server in a production 
 #   environment.
 #
+# @param default_ssl_reload_on_change
```
Thanks for the update of the docs. That actually highlighted something I missed: that it's only the default vhost. Should the name make that explicit?
```suggestion
# @param default_vhost_ssl_reload_on_change
```

timdeluxe replied:
@ekohl Oh, well, right now i see it: All the default_* parameters in init.pp are also used in vhost.pp for non-default vhosts, as long as there is no own parameter defined for that vhost.
If you agree i would remove the complete "note" i've added to the docs, because it is not correct.

ekohl commented on REFERENCE.md:
```diff
@@ -7918,6 +7952,14 @@ Data type: `Any`
 
 Default value: `$apache::params::ssl_certs_dir`
 
+##### <a name="ssl_reload_on_change"></a>`ssl_reload_on_change`
```
Somehow this type has no docs, but I don't see why. The code has a parameter.

timdeluxe replied:
@ekohl I don't get what you mean, i see nothing wrong...?

ekohl replied:
Normally I would expect a description. However, now I don't see any text.

timdeluxe replied:
I still don't get it. Other parameters don't have any text as well in that section. What exactly are you missing?

timdeluxe replied:
@ekohl Can you please give feedback here?

ekohl replied:
I traced it down to https://github.com/puppetlabs/puppetlabs-apache/pull/2165.

timdeluxe replied:
ekohl: Good finding. How should we proceed? Should i merge your commit into my branch here and after that regenerate REFERENCE.md (resulting in 3 commits)? 
Or is it okay, when i do the simple fix by myself and regenerate REFERENCE.md as well and put it into one commit (like i usually did here by force pushes)?

ekohl replied:
It'd be great if you could rebase on my PR. I now regenerated REFERENCE.md in it so your diff should remain small.

ekohl reviewed (commented):
Perhaps some example in https://github.com/puppetlabs/puppetlabs-apache/tree/main/examples would be nice?

timdeluxe commented:
> Perhaps some example in https://github.com/puppetlabs/puppetlabs-apache/tree/main/examples would be nice?

I don't think that this feature is so important, that it needs an example. The ones who need it, will find it anyway...

ekohl reviewed (commented):
Code wise :+1: but I see a few merge commits. Could you rebase?

timdeluxe commented:
> Code wise 👍 but I see a few merge commits. Could you rebase?

After a fighting a bit with git logic (i always get knots in my brain) it is hopefully correctly done now.

ekohl reviewed (approved):
I started CI. I also hope someone else can review but :+1: from me. Thanks for getting this patch in good shape!

timdeluxe commented:
> I started CI. I also hope someone else can review but 👍 from me. Thanks for getting this patch in good shape!

I hope the same. Also thanks to you for guiding me through it and spotting a lot of little things to improve.

timdeluxe commented:
Tests on RedHat failed not due to my code but due to mirrors not answering correctly - don't know if it is a temporary or a general problem, which can be fixed :/

ekohl commented:
That's a known issue (https://github.com/puppetlabs/puppetlabs-apache/pull/2171 tries to look into it)

david22swan reviewed (approved):
LGTM

Will confer with another team member first but look's good to merge.
